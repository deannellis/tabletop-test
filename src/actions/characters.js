import uuid from 'uuid';

// ADD_CHARACTER
export const addCharacter = (
    {
        abilities = {
            strength: 0,
            intelligence: 0,
            wisdom: 0,
            dexterity: 0,
            constitution: 0,
            charisma: 0
        }, 
        race = '', 
        charClass = '', 
        hp = 0, 
        xp = 0,
        name = '',
        description = ''
    } = {}
) => ({
    type: 'ADD_CHARACTER',
    character: {
        id: uuid(),
        abilities,
        race,
        charClass,
        hp,
        xp,
        name,
        description
    }
});

// REMOVE_CHARACTER

export const removeCharacter = ({ id } = {}) => ({
    type: 'REMOVE_CHARACTER',
    id
});

// EDIT_CHARACTER
export const editCharacter = (id, updates) => ({
    type: 'EDIT_CHARACTER',
    id,
    updates
});

