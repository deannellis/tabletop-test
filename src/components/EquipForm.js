import React from 'react';
import { connect } from 'react-redux';
import weapons from '../objects/weapons';
import WeaponListItem from './WeaponListItem';

class EquipForm extends React.Component {
    constructor(props){
        super(props);
        
        const characters = this.props.characters;
        const lastCharacter = characters[characters.length - 1];
        this.currentCharacter = lastCharacter;
        this.state = {
            ids: [],
            gold: lastCharacter.gold
        }
    }
    
    render() {
        
        return (
            <div>
                <form onSubmit={this.onSubmit}>
                    Equip form
                    <h3>{this.state.gold}</h3>
                    {weapons.map((weapon) => {
                        return <WeaponListItem 
                                    key={weapon.id} 
                                    {...weapon}
                                    handleToggleCheckbox={this.handleToggleCheckbox}
                                    currentGold={this.state.gold}
                                />
                    })}
                    <button>Finish</button>
                </form>
            </div>    
        );
    }
    
    onSubmit = (e) => {
        e.preventDefault();
        console.log(this.state.names);
        this.props.onSubmit( this.currentCharacter.id, {
            equipment: this.state.ids
        });
    }
    
    handleToggleCheckbox = (isChecked, id, price) => {
        if(isChecked) {
            this.setState({ gold: this.state.gold - price});
            this.setState({ ids: [ ...this.state.ids, id ]});
        } else {
            this.setState({ gold: this.state.gold + price});
            const array = [...this.state.ids]; // make a separate copy of the array
            const index = array.indexOf(id);
            array.splice(index, 1);
            this.setState({ids: array});
        }
    }
}

const mapStateToProps = (state) => {
    return {
        characters: state.characters
    };
};

export default connect(mapStateToProps)(EquipForm);