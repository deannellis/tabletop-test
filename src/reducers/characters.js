//  CHARACTERS REDUCER
import actionIds from 'actions/actionIds';
const {ADD_CHARACTER, REMOVE_CHARACTER, EDIT_CHARACTER, COMPLETE_CHARACTER} = actionIds;
const charactersReducerDefaultState = [];

export default ( state = charactersReducerDefaultState, action ) => {
    switch (action.type) {
        case ADD_CHARACTER:
            return [
                ...state,
                action.character
            ];
        case REMOVE_CHARACTER:
            return state.filter(({ id }) => {
                return id !== action.id;
            });
        case EDIT_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    return {
                        ...character,
                        ...action.updates
                    }
                } else {
                    return character;
                }
            });
        case COMPLETE_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    let {inProgressStep, ...rest} = character;
                    console.log('rest'+rest);
                    return rest;
                } else {
                    return character;
                }
            });
        default:
            return state;
    }
};