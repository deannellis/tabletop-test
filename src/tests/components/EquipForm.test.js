/* global jest expect */
import React from 'react';
import { shallow } from 'enzyme';
import EquipForm from '../../components/EquipForm';
import characters from '../fixtures/characters';

test('should render equip form without props (character not found)', () => {
    const wrapper = shallow(<EquipForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render equip form with props', () => {
    const wrapper = shallow(<EquipForm currentCharacter={characters[0]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should call onSubmit prop for valid submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<EquipForm currentCharacter={characters[1]} onSubmit={onSubmitSpy} />);
    const value = [ 0, 1, 2 ];
    wrapper.setState({ ids: value });
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(onSubmitSpy).toHaveBeenLastCalledWith(
        characters[1].id,
        {
            equipment: value,
            gold: characters[1].gold
        }
    );
});