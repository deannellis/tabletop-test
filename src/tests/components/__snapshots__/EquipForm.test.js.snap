// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render equip form with props 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    Equip form
    <h3>
      109
    </h3>
    <Component
      currentGold={109}
      damage={6}
      handleToggleCheckbox={[Function]}
      id={0}
      key="0"
      name="Hand Axe"
      price={400}
      range={1}
      size={1}
      weight={5}
    />
    <Component
      currentGold={109}
      damage={8}
      handleToggleCheckbox={[Function]}
      id={1}
      key="1"
      name="Battle Axe"
      price={700}
      range={1}
      size={2}
      weight={7}
    />
    <Component
      currentGold={109}
      damage={10}
      handleToggleCheckbox={[Function]}
      id={2}
      key="2"
      name="Great Axe"
      price={1400}
      range={1}
      size={3}
      weight={10}
    />
    <Component
      currentGold={109}
      damage={10}
      handleToggleCheckbox={[Function]}
      id={3}
      key="3"
      name="Shortbow"
      price={2500}
      range={
        Array [
          70,
          140,
          210,
        ]
      }
      size={2}
      weight={2}
    />
    <button>
      Finish
    </button>
  </form>
</div>
`;

exports[`should render equip form without props (character not found) 1`] = `<NotFoundPage />`;
